Add persistence support for ResourceMonitorThrottler

basuke opened the pull request:
#### 41866885204eefce7d8bfe48862f1a75e970967f
<pre>
Add persistence support for ResourceMonitorThrottler
<a href="https://bugs.webkit.org/show_bug.cgi?id=286652">https://bugs.webkit.org/show_bug.cgi?id=286652</a>
<a href="https://rdar.apple.com/142653861">rdar://142653861</a>

Reviewed by Chris Dumez.

ResourceMonitorThrottler doesn&apos;t have persistent support yet and the stored data will be gone
when application quits. Adding persistence support for ResourceMonitorThrottler to make the data
meaningful over launch to launch.

Also change from ApproximateTime to ContinuousApproximateTime to keep counting while sleeping.

Add tests for persistence on API test.

* Source/WTF/wtf/ContinuousApproximateTime.cpp:
(WTF::ContinuousApproximateTime::fromWallTime):
* Source/WTF/wtf/ContinuousApproximateTime.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/ResourceMonitorPersistence.cpp: Added.
(WebCore::ResourceMonitorPersistence::ResourceMonitorPersistence):
(WebCore::ResourceMonitorPersistence::~ResourceMonitorPersistence):
(WebCore::continuousApproximateTimeToDouble):
(WebCore::doubleToContinuousApproximateTime):
(WebCore::ResourceMonitorPersistence::reportSQLError):
(WebCore::ResourceMonitorPersistence::openDatabase):
(WebCore::ResourceMonitorPersistence::deleteAllRecords):
(WebCore::ResourceMonitorPersistence::deleteExpiredRecords):
(WebCore::ResourceMonitorPersistence::importRecords):
(WebCore::ResourceMonitorPersistence::recordAccess):
(WebCore::ResourceMonitorPersistence::closeDatabase):
* Source/WebCore/loader/ResourceMonitorPersistence.h: Copied from Source/WebCore/loader/ResourceMonitorThrottler.h.
* Source/WebCore/loader/ResourceMonitorThrottler.cpp:
(WebCore::ResourceMonitorThrottler::ResourceMonitorThrottler):
(WebCore::ResourceMonitorThrottler::~ResourceMonitorThrottler):
(WebCore::ResourceMonitorThrottler::throttlerForHost):
(WebCore::ResourceMonitorThrottler::removeOldestThrottler):
(WebCore::ResourceMonitorThrottler::tryAccess):
(WebCore::ResourceMonitorThrottler::clearAllData):
(WebCore::ResourceMonitorThrottler::maintainHosts):
(WebCore::ResourceMonitorThrottler::setCountPerDuration):
(WebCore::ResourceMonitorThrottler::AccessThrottler::tryAccessAndUpdateHistory):
(WebCore::ResourceMonitorThrottler::AccessThrottler::oldestAccessTime const):
(WebCore::ResourceMonitorThrottler::AccessThrottler::tryExpire):
(WebCore::ResourceMonitorThrottler::create): Deleted.
* Source/WebCore/loader/ResourceMonitorThrottler.h:
* Source/WebCore/loader/ResourceMonitorThrottlerHolder.cpp: Added.
(WebCore::ResourceMonitorThrottlerHolder::sharedWorkQueueSingleton):
(WebCore::ResourceMonitorThrottlerHolder::create):
(WebCore::ResourceMonitorThrottlerHolder::ResourceMonitorThrottlerHolder):
(WebCore::ResourceMonitorThrottlerHolder::~ResourceMonitorThrottlerHolder):
(WebCore::ResourceMonitorThrottlerHolder::tryAccess):
(WebCore::ResourceMonitorThrottlerHolder::clearAllData):
(WebCore::ResourceMonitorThrottlerHolder::setCountPerDuration):
* Source/WebCore/loader/ResourceMonitorThrottlerHolder.h: Copied from Source/WebCore/loader/ResourceMonitorThrottler.h.
* Source/WebCore/platform/sql/SQLiteDatabase.cpp:
(WebCore::SQLiteDatabase::indexExists):
* Source/WebCore/platform/sql/SQLiteDatabase.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::shouldOffloadIFrameForHost):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::deleteWebsiteDataImpl):
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::m_resourceMonitorThrottlerDirectory):
(WebKit::NetworkSession::resourceMonitorThrottler):
(WebKit::NetworkSession::protectedResourceMonitorThrottler):
(WebKit::NetworkSession::clearResourceMonitorThrottlerData):
(WebKit::m_isDeclarativeWebPushEnabled): Deleted.
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.serialization.in:
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreConfigurationRef.cpp:
(WKWebsiteDataStoreConfigurationCopyResourceMonitorThrottlerDirectory):
(WKWebsiteDataStoreConfigurationSetResourceMonitorThrottlerDirectory):
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreConfigurationRef.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(-[_WKWebsiteDataStoreConfiguration _resourceMonitorThrottlerDirectory]):
(-[_WKWebsiteDataStoreConfiguration _setResourceMonitorThrottlerDirectory:]):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::defaultResourceMonitorThrottlerDirectory):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::resolveDirectories):
(WebKit::WebsiteDataStore::parameters):
(WebKit::WebsiteDataStore::defaultResourceMonitorThrottlerDirectory):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::initializePaths):
(WebKit::WebsiteDataStoreConfiguration::Directories::isolatedCopy const):
(WebKit::WebsiteDataStoreConfiguration::Directories::isolatedCopy):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::resourceMonitorThrottlerDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setResourceMonitorThrottlerDirectory):
* Tools/TestWebKitAPI/Tests/WebCore/ResourceMonitor.cpp:
(TestWebKitAPI::ResourceMonitorTest::prepareThrottler):
(TestWebKitAPI::ResourceMonitorTest::throttler):
(TestWebKitAPI::ResourceMonitorTest::now):
(TestWebKitAPI::ResourceMonitorTest::later):
(TestWebKitAPI::ResourceMonitorTest::tryAccess):
(TestWebKitAPI::ResourceMonitorTest::temporaryDatabasePath):
(TestWebKitAPI::TEST_F(ResourceMonitorTest, ThrottlerPersistence)):
(TestWebKitAPI::TEST_F(ResourceMonitorTest, ThrottlerCreateAndDelete)):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::configureWebsiteDataStoreTemporaryDirectories):

Canonical link: <a href="https://commits.webkit.org/290576@main">https://commits.webkit.org/290576@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/052528ce7944bdd748d0686e54acaf8b47d3fe2d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90340 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9870 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45265 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95344 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41116 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92392 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10257 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18187 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69538 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27121 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93341 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7852 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81949 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49901 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7579 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36314 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40247 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/83144 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77903 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37384 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97169 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/89118 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17528 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12884 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78534 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17785 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77779 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77754 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19229 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22205 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20813 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10787 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17538 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22865 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/111609 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17279 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26703 "Found 15 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.bytecode-cache, microbenchmarks/memcpy-wasm-medium.js.default, microbenchmarks/memcpy-wasm-medium.js.dfg-eager, microbenchmarks/memcpy-wasm-small.js.dfg-eager, microbenchmarks/memcpy-wasm-small.js.no-llint, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-eager-jettison, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-collect-continuously, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-eager-jettison, wasm.yaml/wasm/stress/ipint-bbq-osr-with-try5.js.default-wasm, wasm.yaml/wasm/stress/ipint-bbq-osr-with-try5.js.wasm-slow-memory ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20731 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19063 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -27,31 +27,55 @@
     ASSERT(maxHosts >= 1);
     RESOURCEMONITOR_RELEASE_LOG("Initialized with count: %zu, duration: %.f, maxHosts: %zu", count, duration.value(), maxHosts);
+
+    protectedWorkQueue()->dispatch([this, path = crossThreadCopy(databasePath)] mutable {
```
In WebKit we avoid having classes with methods that can be called on different threads, as it is too error-prone. I expect you should create a new class for all operations that happen on the work queue, and ResourceMonitorThrottler keeps getting used on the main thread only.

basuke replied:
Got it.

cdumez commented on Source/WebCore/loader/ResourceMonitorPersistence.cpp:
```diff
@@ -0,0 +1,194 @@
+        if (!m_sqliteDB->executeCommandSlow(createIndexSQL(hostIndexName, hostIndexBody)))
+            return reportErrorAndCloseDatabase("create `host` index on `record` table"_s);
+
+        RESOURCEMONITOR_RELEASE_LOG("openDatabase: Index %s created", hostIndexName.characters());
```
Without `PUBLIC_LOG_STRING`, you will not see the index name in external logs.

cdumez commented on Source/WebCore/loader/ResourceMonitorPersistence.cpp:
```diff
@@ -0,0 +1,194 @@
+        if (!m_sqliteDB->executeCommandSlow(createIndexSQL(accessIndexName, accessIndexBody)))
+            return reportErrorAndCloseDatabase("create `access` index on `record` table"_s);
+
+        RESOURCEMONITOR_RELEASE_LOG("openDatabase: Index %s created", accessIndexName.characters());
```
Without `PUBLIC_LOG_STRING`, you will not see the index name in external logs.

cdumez commented on Source/WebCore/loader/ResourceMonitorPersistence.cpp:
```diff
@@ -0,0 +1,194 @@
+
+    Vector<Record> records;
+    while (selectStatement->step() == SQLITE_ROW) {
+        const auto& host = selectStatement->columnText(0);
```
just `auto`

cdumez commented on Source/WebCore/loader/ResourceMonitorPersistence.cpp:
```diff
@@ -0,0 +1,194 @@
+    Vector<Record> records;
+    while (selectStatement->step() == SQLITE_ROW) {
+        const auto& host = selectStatement->columnText(0);
+        const auto& access = selectStatement->columnDouble(1);
```
just `auto`

cdumez commented on Source/WebCore/loader/ResourceMonitorPersistence.cpp:
```diff
@@ -0,0 +1,194 @@
+
+namespace WebCore {
+
+ResourceMonitorPersistence::ResourceMonitorPersistence(String&& path)
```
Every function in this class needs assertions to make sure they are not called on the main thread.

cdumez commented on Source/WebCore/loader/ResourceMonitorPersistence.h:
```diff
@@ -0,0 +1,61 @@
+        ApproximateTime time;
+    };
+
+    WEBCORE_EXPORT ResourceMonitorPersistence(String&& databasePath);
```
`explicit`

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -27,42 +27,80 @@
     ASSERT(maxHosts >= 1);
     RESOURCEMONITOR_RELEASE_LOG("Initialized with count: %zu, duration: %.f, maxHosts: %zu", count, duration.value(), maxHosts);
+
+    protectedWorkQueue()->dispatch([this, path = crossThreadCopy(databasePath)] mutable {
```
Capturing `this` without protection here is unsafe. Nothing guarantees `this` will still be alive when the lambda runs.

call `relaxAdoptionRequirement();` in the constructor if needed to make this happen.

basuke replied:
Ah, right.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -27,42 +27,80 @@
+
+        bool changed = false;
+        for (const auto& record : m_persistence->importRecords()) {
+            auto& throttler = throttlerForHost(record.host);
```
Looks like we're accessing the throttler off the main thread here. Isn't the throttler a main-thread object?

basuke replied:
All throttler objects are touched in worker thread. But I think I forget to release them in the thread which may be the issue.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -27,42 +27,80 @@
+        bool changed = false;
+        for (const auto& record : m_persistence->importRecords()) {
+            auto& throttler = throttlerForHost(record.host);
+            changed |= throttler.tryAccessAndUpdateHistory(record.time, m_config);
```
I suspect some of the throttler functions are called off the main thread and some are not, which is error-prone. Odds are we should be dispatching to the main thread before updating the throttler?

basuke replied:
For early dispatching, it's safe because every change is done in worker thread. We can inform the result without waiting the maintaining the information.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -27,42 +27,80 @@
+
+ResourceMonitorThrottler::~ResourceMonitorThrottler()
+{
+    protectedWorkQueue()->dispatchSync([this] mutable {
```
`this` capture without protection is not safe.

basuke replied:
In deconstructor, `this` won't be invalid during sync call. Do we still need to avoid this pattern?

cdumez replied:
Never mind, this one is sync so it is fine.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -27,42 +27,80 @@
 
 auto ResourceMonitorThrottler::throttlerForHost(const String& host) -> AccessThrottler&
 {
+    ASSERT(!isMainThread());
```
Per previous review, we don't want to have functions called on one thread and others on another thread, on the same class. This is too error-prone.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -77,27 +115,47 @@ void ResourceMonitorThrottler::removeOldestThrottler()
 
-    auto& throttler = throttlerForHost(host);
-    auto result = throttler.tryAccessAndUpdateHistory(time, m_config);
+    protectedWorkQueue()->dispatch([this, host = crossThreadCopy(host), time, completionHandler = WTFMove(completionHandler)] mutable {
```
Capturing `this` without protection is not safe.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -77,27 +115,47 @@ void ResourceMonitorThrottler::removeOldestThrottler()
-    auto result = throttler.tryAccessAndUpdateHistory(time, m_config);
+    protectedWorkQueue()->dispatch([this, host = crossThreadCopy(host), time, completionHandler = WTFMove(completionHandler)] mutable {
+        auto& throttler = throttlerForHost(host);
+        auto result = throttler.tryAccessAndUpdateHistory(time, m_config);
```
I recommend using `bool` instead of `auto` for clarity. It is the same length.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -134,8 +192,10 @@ bool ResourceMonitorThrottler::AccessThrottler::tryExpire(ApproximateTime time,
 {
-    m_config.count = count;
-    m_config.duration = duration;
+    protectedWorkQueue()->dispatch([this, count, duration] mutable {
```
`this` is not safe here.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -2751,4 +2767,13 @@ bool WebsiteDataStore::builtInNotificationsEnabled() const
+WebCore::ResourceMonitorThrottler& WebsiteDataStore::resourceMonitorThrottler()
+{
+    if (!m_resourceMonitorThrottler)
+        m_resourceMonitorThrottler = WTF::makeUnique<WebCore::ResourceMonitorThrottler>(configuration().resourceMonitorThrottlerDirectory());
```
`WTF::` should not be necessary.

bnham commented on Source/WebCore/loader/ResourceMonitorPersistence.cpp:
```diff
@@ -0,0 +1,194 @@
+    return records;
+}
+
+void ResourceMonitorPersistence::recordAccess(const String& host, ApproximateTime time)
```
`ApproximateTime` is not serializable since the epoch resets to zero every time you reboot your device (i.e. `mach_approximate_time()` will be close to 0 every time you reboot). The same goes with `MonotonicTime`.

Only wall clock timestamps can be persisted. You can use `approximateWallTime()` to convert the `ApproximateTime` to a `WallTime`.

cdumez commented on Source/WebCore/loader/ResourceMonitorPersistence.cpp:
```diff
@@ -0,0 +1,205 @@
+static ContinuousApproximateTime doubleToContinuousApproximateTime(double timestamp)
+{
+    auto wallTime = WallTime::fromRawSeconds(timestamp);
+    return ContinuousApproximateTime::now() + (wallTime - WallTime::now());
```
Ideally we'd have a `fromWallTime()` static function on `ContinuousApproximateTime`

cdumez commented on Source/WebCore/loader/ResourceMonitorPersistence.cpp:
```diff
@@ -0,0 +1,205 @@
+
+ResourceMonitorPersistence::~ResourceMonitorPersistence()
+{
+    if (m_sqliteDB)
```
Needs an assertion to make sure it runs off the main thread.

cdumez commented on Source/WebCore/loader/ResourceMonitorPersistence.cpp:
```diff
@@ -0,0 +1,205 @@
+
+namespace WebCore {
+
+ResourceMonitorPersistence::ResourceMonitorPersistence() = default;
```
Needs an assertion to make sure it runs off the main thread.

cdumez commented on Source/WebCore/loader/ResourceMonitorPersistence.h:
```diff
@@ -0,0 +1,64 @@
+    WEBCORE_EXPORT ~ResourceMonitorPersistence();
+
+    bool openDatabase(String&& path);
+    bool isDatabaseOpened() const { return m_sqliteDB && m_sqliteDB->isOpen(); }
```
`isDatabaseOpen()` sounds more correct.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -27,47 +27,147 @@
+    ASSERT(isMainThread());
+
+    sharedWorkQueueSingleton()->dispatch([container = WTFMove(m_container)] mutable {
+        container.reset(nullptr);
```
This is unnecessary, the dispatch and lambda capture is sufficient.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -27,47 +27,147 @@
+
+    sharedWorkQueueSingleton()->dispatch([weakThis = ThreadSafeWeakPtr { *this }, host = crossThreadCopy(host), time, completionHandler = WTFMove(completionHandler)] mutable {
+        RefPtr protectedThis = weakThis.get();
+        bool wasGranted = protectedThis ? protectedThis->m_container->tryAccess(host, time) : false;
```
```
bool wasGranted = protectedThis && protectedThis->m_container->tryAccess(host, time);
```

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -27,47 +27,147 @@
+        RefPtr protectedThis = weakThis.get();
+        bool wasGranted = protectedThis ? protectedThis->m_container->tryAccess(host, time) : false;
+
+        callOnMainRunLoop([wasGranted, completionHandler = WTFMove(completionHandler)] mutable {
```
Should be `callOnMainThread()` in WebCore/

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -27,47 +27,147 @@
+        if (RefPtr protectedThis = weakThis.get())
+            protectedThis->m_container->clearAllData();
+
+        callOnMainRunLoop(WTFMove(completionHandler));
```
`callOnMainThread`

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -27,47 +27,147 @@
+    });
+}
+
+ResourceMonitorThrottler::ThrottlerContainer::ThrottlerContainer(String&& path, size_t count, Seconds duration, size_t maxHosts)
```
This is used on a non main thread, please move to another file.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.cpp:
```diff
@@ -81,31 +181,60 @@ void ResourceMonitorThrottler::removeOldestThrottler()
+        m_persistence->deleteAllRecords();
+}
+
+void ResourceMonitorThrottler::ThrottlerContainer::maintainHosts(ContinuousApproximateTime time)
```
All this `ThrottlerContainer` class should be renamed and moved to its own file. We should never have functions called on various threads in the same file / class.

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.h:
```diff
@@ -25,24 +25,38 @@
-    WTF_MAKE_FAST_ALLOCATED;
+class ResourceMonitorPersistence;
+
+class ResourceMonitorThrottler final : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<ResourceMonitorThrottler, WTF::DestructionThread::MainRunLoop> {
```
`MainThread` in WebCore/ not `MainRunLoop`

cdumez commented on Source/WebCore/loader/ResourceMonitorThrottler.h:
```diff
@@ -51,29 +65,62 @@ class ResourceMonitorThrottler final : public RefCounted<ResourceMonitorThrottle
 
-    ResourceMonitorThrottler(size_t count, Seconds duration, size_t maxHosts);
+    // Shared WorkQueue is used to prevent race condition when delete and create Throttler for same database file.
+    static Ref<WorkQueue> sharedWorkQueueSingleton()
```
This should not be inlined in the header.

basuke committed:
Add persistence support for ResourceMonitorThrottler
https://bugs.webkit.org/show_bug.cgi?id=286652
rdar://142653861

Reviewed by Chris Dumez.

ResourceMonitorThrottler doesn't have persistent support yet and the stored data will be gone
when application quits. Adding persistence support for ResourceMonitorThrottler to make the data
meaningful over launch to launch.

Also change from ApproximateTime to ContinuousApproximateTime to keep counting while sleeping.

Add tests for persistence on API test.

* Source/WTF/wtf/ContinuousApproximateTime.cpp:
(WTF::ContinuousApproximateTime::fromWallTime):
* Source/WTF/wtf/ContinuousApproximateTime.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/ResourceMonitorPersistence.cpp: Added.
(WebCore::ResourceMonitorPersistence::ResourceMonitorPersistence):
(WebCore::ResourceMonitorPersistence::~ResourceMonitorPersistence):
(WebCore::continuousApproximateTimeToDouble):
(WebCore::doubleToContinuousApproximateTime):
(WebCore::ResourceMonitorPersistence::reportSQLError):
(WebCore::ResourceMonitorPersistence::openDatabase):
(WebCore::ResourceMonitorPersistence::deleteAllRecords):
(WebCore::ResourceMonitorPersistence::deleteExpiredRecords):
(WebCore::ResourceMonitorPersistence::importRecords):
(WebCore::ResourceMonitorPersistence::recordAccess):
(WebCore::ResourceMonitorPersistence::closeDatabase):
* Source/WebCore/loader/ResourceMonitorPersistence.h: Copied from Source/WebCore/loader/ResourceMonitorThrottler.h.
* Source/WebCore/loader/ResourceMonitorThrottler.cpp:
(WebCore::ResourceMonitorThrottler::ResourceMonitorThrottler):
(WebCore::ResourceMonitorThrottler::~ResourceMonitorThrottler):
(WebCore::ResourceMonitorThrottler::throttlerForHost):
(WebCore::ResourceMonitorThrottler::removeOldestThrottler):
(WebCore::ResourceMonitorThrottler::tryAccess):
(WebCore::ResourceMonitorThrottler::clearAllData):
(WebCore::ResourceMonitorThrottler::maintainHosts):
(WebCore::ResourceMonitorThrottler::setCountPerDuration):
(WebCore::ResourceMonitorThrottler::AccessThrottler::tryAccessAndUpdateHistory):
(WebCore::ResourceMonitorThrottler::AccessThrottler::oldestAccessTime const):
(WebCore::ResourceMonitorThrottler::AccessThrottler::tryExpire):
(WebCore::ResourceMonitorThrottler::create): Deleted.
* Source/WebCore/loader/ResourceMonitorThrottler.h:
* Source/WebCore/loader/ResourceMonitorThrottlerHolder.cpp: Added.
(WebCore::ResourceMonitorThrottlerHolder::sharedWorkQueueSingleton):
(WebCore::ResourceMonitorThrottlerHolder::create):
(WebCore::ResourceMonitorThrottlerHolder::ResourceMonitorThrottlerHolder):
(WebCore::ResourceMonitorThrottlerHolder::~ResourceMonitorThrottlerHolder):
(WebCore::ResourceMonitorThrottlerHolder::tryAccess):
(WebCore::ResourceMonitorThrottlerHolder::clearAllData):
(WebCore::ResourceMonitorThrottlerHolder::setCountPerDuration):
* Source/WebCore/loader/ResourceMonitorThrottlerHolder.h: Copied from Source/WebCore/loader/ResourceMonitorThrottler.h.
* Source/WebCore/platform/sql/SQLiteDatabase.cpp:
(WebCore::SQLiteDatabase::indexExists):
* Source/WebCore/platform/sql/SQLiteDatabase.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::shouldOffloadIFrameForHost):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::deleteWebsiteDataImpl):
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::m_resourceMonitorThrottlerDirectory):
(WebKit::NetworkSession::resourceMonitorThrottler):
(WebKit::NetworkSession::protectedResourceMonitorThrottler):
(WebKit::NetworkSession::clearResourceMonitorThrottlerData):
(WebKit::m_isDeclarativeWebPushEnabled): Deleted.
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.serialization.in:
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreConfigurationRef.cpp:
(WKWebsiteDataStoreConfigurationCopyResourceMonitorThrottlerDirectory):
(WKWebsiteDataStoreConfigurationSetResourceMonitorThrottlerDirectory):
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreConfigurationRef.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(-[_WKWebsiteDataStoreConfiguration _resourceMonitorThrottlerDirectory]):
(-[_WKWebsiteDataStoreConfiguration _setResourceMonitorThrottlerDirectory:]):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::defaultResourceMonitorThrottlerDirectory):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::resolveDirectories):
(WebKit::WebsiteDataStore::parameters):
(WebKit::WebsiteDataStore::defaultResourceMonitorThrottlerDirectory):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::initializePaths):
(WebKit::WebsiteDataStoreConfiguration::Directories::isolatedCopy const):
(WebKit::WebsiteDataStoreConfiguration::Directories::isolatedCopy):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::resourceMonitorThrottlerDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setResourceMonitorThrottlerDirectory):
* Tools/TestWebKitAPI/Tests/WebCore/ResourceMonitor.cpp:
(TestWebKitAPI::ResourceMonitorTest::prepareThrottler):
(TestWebKitAPI::ResourceMonitorTest::throttler):
(TestWebKitAPI::ResourceMonitorTest::now):
(TestWebKitAPI::ResourceMonitorTest::later):
(TestWebKitAPI::ResourceMonitorTest::tryAccess):
(TestWebKitAPI::ResourceMonitorTest::temporaryDatabasePath):
(TestWebKitAPI::TEST_F(ResourceMonitorTest, ThrottlerPersistence)):
(TestWebKitAPI::TEST_F(ResourceMonitorTest, ThrottlerCreateAndDelete)):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::configureWebsiteDataStoreTemporaryDirectories):

Canonical link: https://commits.webkit.org/290576@main
